zerver/lib/bugdown/__init__.py

```diff
@@ -1504,12 +1504,17 @@ def maybe_update_markdown_engines(realm_filters_key, email_gateway):
     else:
         realm_filters = realm_filters_for_realm(realm_filters_key)
         if realm_filters_key not in realm_filter_data or    \
-                realm_filter_data[realm_filters_key] != realm_filters or    \
-                (realm_filters_key, email_gateway) not in md_engines:
-            # Either realm filters data has changed or an email message has been
-            # forwarded to an realm by the email gateway for which the markdown
-            # engine specific to the email gateway hasn't been populated.
+                realm_filter_data[realm_filters_key] != realm_filters:
+            # Realm filters data has changed, update `realm_filter_data` and any
+            # of the existing markdown engines using this set of realm filters.
             realm_filter_data[realm_filters_key] = realm_filters
+            for email_gateway_flag in [True, False]:
+                if (realm_filters_key, email_gateway_flag) in md_engines:
+                    # Update only existing engines(if any), don't create new one.
+                    make_md_engine(realm_filters_key, email_gateway_flag)
```

Comment:
I don't like this fix, because it means we end up constructing the markdown engines for both values of `email_gateway_flag`, even though `email_gateway=True` is very rare.  Since constructing these engines is somewhat expensive both in RAM and CPU, I don't think we want to do that.  That's why I was suggesting just making the key that we save the state under include the `email_gateway` flag...

Reply:
Yeah I know that it is expensive to create markdown engines and the fix doesn't create any new engines just updates the existing one. As far as I can see this is going to have same effect as saving the state using `email_gateway` flag. Can you kindly go through the commit once more?

Reply:
Oh, I see, I'd missed the `if` statement here.  OK.